### PR TITLE
adding function existsPath together with test

### PR DESCRIPTION
**Pull request by pnrobinson**

### CODE REVIEW

File: ontolib-core/src/main/java/com/github/phenomics/ontolib/ontology/data/ImmutableOntology.java [hunk 1]
```diff
@@ -1,9 +1,6 @@
 package com.github.phenomics.ontolib.ontology.data;
 
-import java.util.Collection;
```
Comment (holtgrewe) [inline ontolib-core/src/main/java/com/github/phenomics/ontolib/ontology/data/ImmutableOntology.java:3]: Please no star imports but explicit ones. I guess you can easily adjust your IDE to do this.

File: ontolib-core/src/main/java/com/github/phenomics/ontolib/ontology/data/ImmutableOntology.java [hunk 2]
```diff
@@ -188,6 +185,25 @@ public TermId getRootTermId() {
     return obsoleteTermIds;
   }
 
+  @Override
+  public boolean existsPath(final TermId sourceID, TermId destID){
+    // special case -- a term cannot have a path to itself in an ontology (DAG)
+    if (sourceID.equals(destID)) return false;
```
Comment (holtgrewe) [inline ontolib-core/src/main/java/com/github/phenomics/ontolib/ontology/data/ImmutableOntology.java:191]: Please wrap and indent consistently. I can send around a style file for Eclipse.

File: ontolib-core/src/main/java/com/github/phenomics/ontolib/ontology/data/ImmutableOntology.java [hunk 3]
```diff
@@ -188,6 +185,25 @@ public TermId getRootTermId() {
...
+    BreadthFirstSearch<TermId, ImmutableEdge<TermId>> bfs = new BreadthFirstSearch<>();
+    bfs.startFromForward(graph, sourceID, new VertexVisitor<TermId, ImmutableEdge<TermId>>() {
+        @Override
+        public boolean visit(DirectedGraph<TermId, ImmutableEdge<TermId>> g, TermId termId) {
+          visited.add(termId);
+          return true;
+        }});
+    return visited.contains(destID);
+  }
+
```
Comment (holtgrewe) [inline ontolib-core/src/main/java/com/github/phenomics/ontolib/ontology/data/ImmutableOntology.java:202]: Remove superflous whitespace for consistencty.

File: ontolib-core/src/main/java/com/github/phenomics/ontolib/ontology/data/Ontology.java [hunk 4]
```diff
@@ -97,6 +97,9 @@ default TermId getPrimaryTermId(TermId termId) {
     return result;
   }
 
+  public boolean existsPath(final TermId sourceID, TermId destID);
+
```
Comment (holtgrewe) [inline ontolib-core/src/main/java/com/github/phenomics/ontolib/ontology/data/Ontology.java:101]: Please add javadoc and remove superflous empty lines.

File: ontolib-core/src/test/java/com/github/phenomics/ontolib/ontology/data/ImmutableOntologyTest.java [hunk 5]
```diff
@@ -36,4 +37,26 @@ public void test() {
         ontology.getParentTermIds(id1).toString());
   }
 
+
+  /** The example graph has id1->id2, id1->id3, id1->id4, id2->id5, id4->id5 */
+  @Test
+  public void testPathExists() {
+    assertTrue(ontology.existsPath(id1,id2));
```
Comment (holtgrewe) [inline ontolib-core/src/test/java/com/github/phenomics/ontolib/ontology/data/ImmutableOntologyTest.java:44]: Please make whitespace consistent, your IDE should offer an option to do this automatically.

File: ontolib-core/src/test/java/com/github/phenomics/ontolib/ontology/data/ImmutableOntologyTest.java [hunk 6]
```diff
@@ -36,4 +37,26 @@ public void test() {
...
+    assertFalse(ontology.existsPath(id4,id1));
+    assertTrue(ontology.existsPath(id1,id5));
+    assertFalse(ontology.existsPath(id5,id1));
+    assertTrue(ontology.existsPath(id2,id5));
+    assertFalse(ontology.existsPath(id5,id2));
+    assertTrue(ontology.existsPath(id4,id5));
+    assertFalse(ontology.existsPath(id5,id4));
+  // test that a term cannot have a path to itself.
+    assertFalse(ontology.existsPath(id5,id5));
+
```
Comment (holtgrewe) [inline ontolib-core/src/test/java/com/github/phenomics/ontolib/ontology/data/ImmutableOntologyTest.java:58]: Please remove empty line here and below.

**Review by holtgrewe — CHANGES_REQUESTED** [review]
Overall looks good, see inline annotations.

Also, I'd rather have this in algo.OntologyTerms to keep the `Ontology` interfaces slim and rather have utility classes with static functions.

File: ontolib-core/src/main/java/com/github/phenomics/ontolib/ontology/data/ImmutableOntology.java (hunk 1 above)
Comment (holtgrewe) [reply ontolib-core/src/main/java/com/github/phenomics/ontolib/ontology/data/ImmutableOntology.java:3]: Also, I forgot: please add note to Changelog.

**Comment by pnrobinson** [comment]
I have replaced this PR with PR #39 and will close this

**Pull request closed without merging**